File: Ruby/problem3.rb
load 'problem.rb'
require 'set'

class Problem3 < Problem

  def initialize(list, exp)
    @list = list
    @exp = exp
  end

  def Calculate()
    counter = 0
    s1 = Set[]
    for i in @list
      if i == counter+1
        counter = counter + 1
        while (s1.include?(counter+1))
          counter = counter + 1
        end
      else
        s1.add(i)
      end

    end

    return @exp == counter+1

  end

end

File: Dart/lib/Problem3.dart
import 'package:Dart/Problem.dart' as Problem;

class Problem3 implements Problem.Problem {

  List<int> list;
  int exp;


  Problem3(List<int> list, int exp){
    this.list = list;
    this.exp = exp;
  }


  @override
  bool Calculate() {

    int counter = 0;
    Set<int> set = [].toSet();
    for (int x in this.list){
      if (x == counter+1){
        counter++;
        while (set.contains(counter+1)){
          counter++;
        }
      }else{
        set.add(x);
      }
    }


    return counter+1 == this.exp;

  }
}